TC azimuthal rain rate diagnostic draft code

GNellyEmlaw opened the pull request:
**Description**
The TC Rain POD has been added, with a directory titled "TC_Rain" in the diagnostics subdirectory. The driver script is the "TC_rain.py", which includes comments throughout the script. I am not very familiar the with the os (operating system) module in python, so I imported the data files with local paths, but I will meet with someone in my department to fix that shortly. I just wanted to get the general script into a draft PR so that the developers can make comments on the other parts of the my code more swiftly

Associated issue # (replace this phrase and parentheses with the issue number)  

**How Has This Been Tested?**
The script can be tested by running the TC_rain.py file after changing the paths of the model data and TC track data on lines 81 and 106. I apologize for the inconvenience and will work on implementing os into the code over the next week. 

**Checklist:**
- [x] I have reviewed my own code to ensure that if follows the [POD development guidelines](https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/dev_guidelines.html)
- [x] My branch is up-to-date with the NOAA-GFDL main branch, and all merge conflicts are resolved
- [x] The scripts are written in Python 3.7 or above (preferred; required if funded by a CPO grant), NCL, or R
- [x ] All of my scripts are in the diagnostics/[POD short name] subdirectory, and include a main_driver script, template html, and settings.jsonc file
- [x] I have made corresponding changes to the documentation in the POD's doc/ subdirectory
- [x] I have requested that the framework developers add packages required by my POD to the python3, NCL, or R environment yaml file if necessary, and my environment builds with `conda_env_setup.sh` 
- [ ] I have added any necessary data to input_data/obs_data/[pod short name] and/or input_data/model/[pod short name]
- [x] My code is portable; it uses [MDTF environment variables](https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/ref_envvars.html), and does not contain hard-coded file or directory paths
- [ ] I have provided the code to generate digested data files from raw data files
- [x] Each digested data file generated by the script contains numerical data (no figures), and is 3 GB or less in size
- [x] The repository contains no extra test scripts or data files

GNellyEmlaw commented:
I have pushed my TC Rain branch of the repository so it is up to date and am attaching the log files for both when I try to run the default_tests.jsonc and my TC_Rain_runtime.jsonc file. In both cases there is an error finding the data. 


[QBOi.EXP1.AMIP.001.log](https://github.com/NOAA-GFDL/MDTF-diagnostics/files/9105444/QBOi.EXP1.AMIP.001.log)
[TC_Rain.data.log](https://github.com/NOAA-GFDL/MDTF-diagnostics/files/9105441/TC_Rain.data.log)

GNellyEmlaw commented:
> * copies
@wrongkindofdoctor 

Okay, I was able to redownload the data for the default_tests.jsonc run and was able to sucessfully run that, thank you!

I am still having some trouble with running my POD though. (Sorry, I meant to include the log that wasn't .data for that - attaching it here now) Should I be including my data in both the model folder and the obs_data folder as a dummy file? Previously I had been including it only in the obs folder, but even when I add it to the model directory under the folder name ECMWF, I am still receiving an error. I will attach my file structure as a screen shot from visual studio code and have pushed everything in TC Rain so it is up to date.


[ECMWF.log](https://github.com/NOAA-GFDL/MDTF-diagnostics/files/9115748/ECMWF.log)
![Screen Shot 2022-07-14 at 3 05 00 PM](https://user-images.githubusercontent.com/107883638/179096573-3bad9364-6426-4ca5-8d46-8b9910a0eec9.png)

wrongkindofdoctor commented:
@GNellyEmlaw Okay--good that the CESM tests worked, and your directory paths are defined to match the configuration file settings.

For your POD, you should only need a copy of the data in the model/ECMWF directory. The obs_data/TC_Rain directory can be empty.

In your runtime jsonc, the ["convention" needs to be CMIP](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/b98e979eca5cbece20de1e890fef332d3ebe25ce/diagnostics/TC_Rain%20/TC_Rain_runtime.jsonc#L22) (we also support GFDL and CESM/NCAR data formats, but ECMWF data will most likely adhere to CMIP standards).  

In the POD settings file, the [standard name in varlist section ](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/b98e979eca5cbece20de1e890fef332d3ebe25ce/diagnostics/TC_Rain%20/settings.jsonc#L66)"total precipitation" does not fit any of the definitions in the CMIP fieldlist, and more likely matches the long name metadata entry; the closest match is the MDTF-diagnostics CMIP fieldlist is  [precipitation_flux](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/a6449a79220b5deed781a0ac60610547057678c9/data/fieldlist_CMIP.jsonc#L92-L101). There is no entry for total precipitation at the moment, and no standard definition in the CMIP6 database I could find, so I'll have to add one.

If you do an ncdump on your input file,  the standard name should be defined using underscores (e.g., something like "total_precipitation"). Let me know what that standard name is (and other metadata results are) for the "tp" variable, and I will add an entry for total precip to the CMIP fieldlist. Then, you can update your POD branch by pulling the changes from the main branch.

wrongkindofdoctor commented:
@GNellyEmlaw  Okay, the framework needs a `standard_name` attribute in a model input dataset to reference for the preprocessor to convert it to the proper convention. I just added a variable called ["total_precipitation" to the CMIP fieldlist ](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/data/fieldlist_CMIP.jsonc#L102-L106), and will work on getting the framework to handle long_name as an alternative attribute to use for preprocessing, as well as slight mods to the units (e.g., kg ms-2 vs kg ms**-2 in your data)

As a workaround for your testing in the meantime, you have a couple of options:

* pull in changes from the NOAA_GFDL main branch to your POD branch so your CMIP fieldlist is updated. Then, modify your input data file to append an attribute to the tp variable called `standard_name` with a value of "total_precipitation", and modify your settings file to include the varlist entry standard_name with a value of total_precipitation. You'll also want to change the units   from "kg m**-2" to "kg m-2" in the input file and settings file.

* Bypass the preprocessor entirely by adding the flag `--disable-preprocessor` to the `./mdtf` command: `./mdtf -f [path to runtime file] --disable-preprocessor -v`. In this case, the POD will have to check that the data are correct. You'll want to verify the units and dimensions, and loop through multiple metadata attributes like standard_name and long_name to check for the correct variable name. I recommend making string searches case-insensitive, and accounting for slight variations in the units. 

Regardless of the option you choose, you'll have to generalize your POD driver script to read any input file, not just files from your test ECMWF data. The [example_diag.py](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/main/diagnostics/example/example_diag.py) driver script demonstrates how to do this by [defining a pod environment variable called [[VARIABLE NAME]_FILE](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/example/example_diag.py#L86-L93). 
The framework creates this environment variable for each variable input data file, so your pod script should be able to reference an environment variable called 'TP_FILE' . 

Similarly, you can reference input variable names with the framework generated environment [variable [variable name]_var](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/example/example_diag.py#L102-L104) (all lowercase; your POD would use "tp_var").

GNellyEmlaw commented:
@wrongkindofdoctor 
Okay, I was able to generalize the POD driver script by calling the variable environment with TP_FILE and TC_Rain_runtime.jsonc is running successfully and the package is exiting normally, but I have not been able to save figures I would like output. The line I am using the driver script is:

output_fname =  os.path.join(os.environ.get("WK_DIR"), "model", fname)
plt.savefig(output_fname, format = 'png' )

I have pushed everything so it is all up to date. Is there anything I am missing?

wrongkindofdoctor commented:
@GNellyEmlaw The directory path is correct for files that are saved directly as .png (i.e., if fname ends in .png, you don't need to specify the format)

 If 'fname' ends in .eps as in your current PR commit, the output file path should be:
```
${WK_DIR}/model/PS/[fname]
```
The framework will convert .eps files in the model/PS directory  to .png format so they can be viewed in the html output.
If your're still having trouble, send me the logs [CASENAME].log, and POD NAME].log (in the CASENAME/POD_NAME output directory).

Also, please provide a copy of the test data when you are able so I can work on adding support for the metadata. You can send me a link to a tarball with the output hosted on an ftp server that I can access  via curl, wget, sftp,etc..., or use Globus file transfer if your system has an endpoint (share the directory with jessica.liptak@rdhpcs.noaa.gov).

As an aside, you seem to have 2 TC_rain directories in your diagnostics folder, so you'll want to delete one of them (probably the older one, but double check). I recommend using the web interface to fix this on your remote branch first, then updating your local branch. 
1. In your local copy of your repo, commit any changes that you want to keep, and push them to your remote repo. Also double check that there are also 2 copies of TC_rain in your diagnostics folder on your machine; if there is only one, skip the following steps and let me know.
2. From the web interface in your remote MDTF-diagnostics fork, select the TC_rain branch from the menu
3. Navigate to the directory that you want to delete, then select the [...] button next to the "add file" dropdown menu toward the upper right, and click "delete directory". Scroll down, and commit the change to your branch. 
4. In your local copy of your repo, verify that you are on your pod branch, run `git fetch`, then `git pull` to pull the updates from your remote branch. Verify that you have only one copy of TC_rain your local diagnostics directory.

wrongkindofdoctor commented:
@GNellyEmlaw Thanks--I got the email with your data. If the duplicated directories are only located in the remote repo, don't worry about it. As long as at least one shows changes you push, I can remove the extra POD directory after merging it into the GFDL repo

As for your POD error, your log show that there's a problem with the input file path definition:
```
Traceback (most recent call last):
  File "/home/disk/p/gnemlaw/mdtf/MDTF-diagnostics/diagnostics/TC_Rain/TC_rain.py", line 81, in <module>
    pr_fname = os.path.join(os.environ.get("MODEL_DATA_ROOT"), "ECMWF","day", "*nc")
  File "/home/disk/p/gnemlaw/anaconda3/envs/_MDTF_python3_base/lib/python3.7/posixpath.py", line 80, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```
It looks like the error is because the file name is incorrectly specified, and should follow the format: <CASENAME>.<variable_name>.<frequency>.nc

That said, you should use the framework-generated environment variables instead of the hardcoded path def with "ECMWF" so your POD can work with different datasets. You can use the environment variable for the input file path generated for PR:

```
pr_fname = os.environ["PR_FILE"]
```
If this doesn't work, or you need/prefer the os.path.join method:
```
os.path.join(os.environ["DATADIR"],"day",os.environ["CASENAME"]+"."+os.environ["pr_var"]+".day.nc")
```

GNellyEmlaw commented:
@wrongkindofdoctor 
Thanks for that check, I accidentally changed my track data to os.environ["PR_FILE"] and not my tp data, and was able to get the data running. I think it is accessing the data correctly now, however now the test.jsonc file is not successfully running. It does not finish and I do not get an output/log file. 

It looks like it is having some trouble writing the output file? I am getting the error:

`(base) 113 p% ./mdtf -f ./diagnostics/TC_Rain/TC_Rain_runtime.jsonc
=== Starting /home/disk/p/gnemlaw/mdtf/MDTF-diagnostics/mdtf_framework.py

PACKAGE SETTINGS:
  case_list(0): 
    CASENAME: ECMWF
    model: ECMWF
    convention: CMIP
    FIRSTYR: 2002
    LASTYR: 2002
    pod_list: 
      TC_Rain
    experiment: ""
  paths: 
    CODE_ROOT: /home/disk/p/gnemlaw/mdtf/MDTF-diagnostics
    OBS_DATA_ROOT: /home/disk/p/gnemlaw/mdtf/inputdata/obs_data
    MODEL_DATA_ROOT: /home/disk/p/gnemlaw/mdtf/inputdata/model
    WORKING_DIR: /home/disk/p/gnemlaw/mdtf/wkdir
    OUTPUT_DIR: /home/disk/p/gnemlaw/mdtf/wkdir
    conda_root: /home/disk/p/gnemlaw/anaconda3
    conda_env_root: /home/disk/p/gnemlaw/anaconda3/envs/
    TEMP_DIR_ROOT: /home/disk/p/gnemlaw/mdtf/wkdir
  settings: 
    file_transfer_timeout: 300
    output_manager: html
    save_ps: true
    make_variab_tar: true
    convention: CMIP
    data_manager: Local_File
    environment_manager: Conda
    runtime_manager: subprocess
  env_vars: 
    RGB: /home/disk/p/gnemlaw/mdtf/MDTF-diagnostics/shared/rgb
    MPLBACKEND: Agg
### MDTFFramework: initializing case 'ECMWF'.
### MDTFFramework: requesting data for case 'ECMWF'.
Starting data file search at /home/disk/p/gnemlaw/mdtf/inputdata/model:
Directory crawl found 21 files.
Querying <tp> (total_precipitation).
Fetching <#cmzC:TC_Rain.tp> (='tp' @ 6hr).
Preprocessing <#cmzC:TC_Rain.tp> (='tp' @ 6hr).
Requested dates for <#cmzC:TC_Rain.tp> coincide with range of dataset '2002-07-12 -- 2002-08-27'; left unmodified.
Converted units on <#cmzC:TC_Rain.tp>.
Writing 20 mb to $WK_DIR/TC_Rain/6hr/ECMWF.tp.6hr.nc
WARNING: Conflict in '_FillValue' attribute of longitude: nan != nan.
WARNING: Conflict in '_FillValue' attribute of latitude: nan != nan.
### MDTFFramework: running case 'ECMWF'.
SubprocessRuntimeManager: run <#ccW6:ECMWF.TC_Rain>.
### Starting <#ccW6:ECMWF.TC_Rain>
<#ccW6:ECMWF.TC_Rain> will run using 'python' from conda env '_MDTF_python3_base'.
        Running TC_rain.py for <#ccW6:ECMWF.TC_Rain>.
<#ccW6:ECMWF.TC_Rain> was terminated or exited abnormally.
SubprocessRuntimeManager: completed all PODs.
CRITICAL: **********************************************************************
Uncaught exception:
Traceback (most recent call last):
  File "/home/disk/p/gnemlaw/mdtf/MDTF-diagnostics/mdtf_framework.py", line 67, in <module>
    exit_code = main(argv)
  File "/home/disk/p/gnemlaw/mdtf/MDTF-diagnostics/mdtf_framework.py", line 62, in main
    exit_code = framework.main()
  File "/home/disk/p/gnemlaw/mdtf/MDTF-diagnostics/src/core.py", line 1117, in main
    out_mgr.make_output()
  File "/home/disk/p/gnemlaw/mdtf/MDTF-diagnostics/src/output_manager.py", line 425, in make_output
    open(self.CASE_TEMP_HTML, 'w').close()
FileNotFoundError: [Errno 2] No such file or directory: '/home/disk/p/gnemlaw/mdtf/wkdir/MDTF_ECMWF_2002_2002/_MDTF_pod_output_temp.html'`

GNellyEmlaw commented:
@wrongkindofdoctor 

Thank you, I was able to get the package running after fixing this! Sorry to have needed so much help, I really appreciate you taking the time.

I've pushed, so the TC Rain Branch is up to date. I have a feeling there might be more edits needed to generalize the code. For example, in its current from the the TC_Rain.py code only takes TC tracks from one basin at a time because the data I am working with splits the tracks by basin. On line 73 in the TC_Rain.py file I specify that the basin in the example file is the western north pacific. I have also set the windspeed threshold TCs used in the azimuthal average plotted to be 30 to 40 knots and specify that in on line 74. I imagine these are things that end users may want to change. Is it okay to leave them there and have end users edit that? Or would there be a way to include that in a settings file?

Thanks so much again.

wrongkindofdoctor commented:
@GNellyEmlaw It doesn't look like your updates are being pushed to your remote branch--the most recent commit is from 5 days ago, and this may be why there are still two TC_rain directories (one with a trailing space at the end) in the PR.

If you run `git status`, what is the output?

GNellyEmlaw commented:
@wrongkindofdoctor 

Okay, I realized I was on the wrong branch, but I changed that and it still seems to not be updating to the TC_Rain Branch. 

```
(base) 113 p% git status
On branch TC_Rain
Your branch is up to date with 'origin/TC_Rain'.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   data/fieldlist_CMIP.jsonc
        modified:   data/fieldlist_NCAR.jsonc
        deleted:    diagnostics/TC_Rain /TC_Rain.html
        deleted:    diagnostics/TC_Rain /TC_Rain_runtime.jsonc
        deleted:    diagnostics/TC_Rain /TC_rain.py
        deleted:    diagnostics/TC_Rain /doc/TC_Rain.rst
        deleted:    diagnostics/TC_Rain /settings.jsonc
        modified:   diagnostics/TC_Rain/TC_Rain.html
        modified:   diagnostics/TC_Rain/TC_Rain_runtime.jsonc
        modified:   diagnostics/TC_Rain/TC_rain.py
        modified:   diagnostics/TC_Rain/settings.jsonc
        modified:   src/default_tests.jsonc

no changes added to commit (use "git add" and/or "git commit -a")
(base) 114 p% git push
Username for 'https://github.com': GNellyEmlaw
Password for 'https://GNellyEmlaw@github.com': 
Everything up-to-date
(base) 115 p% 
```

wrongkindofdoctor commented:
@GNellyEmlaw Ah, you haven't committed any of your local changes. Do the following to update your remote branch with changes you made to the files in your POD directory (PR will update automatically):
```
git add diagnostics/TC_rain/*.jsonc
git add diagnostics/TC_rain/TC_rain.py
git add diagnostics/TC_rain/TC_rain.html
```
This will add the changes you made to the files in the TC_rain directory without the trailing blank space.

Now, to remove the duplicate directory with the trailing space from the git commit history, run
```
git rm -r 'diagnostics/TC_rain ' [make sure to add a the space after TC_rain]
```

If you run git status, you should see a message indicating that the changes in the files you just added and removed are staged for a commit. You can now push the changes to your remote branch:
```
git push -u origin TC_rain
```
Once the PR updates and passes the CI, you can move the branch from draft mode to "ready for review". THe  framework team can give a more thorough review of the code, and we make suggestions on how to improve the driver script, including modifying some of the parameters so that users can adjust them. 

This is probably an instance where you'd want to incorporate your own POD environment variables ([pod_env_vars](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/main/doc/sphinx/ref_settings.rst#settings-section)), which are set in the settings.jsonc file as you mentioned. The example POD demonstrates this by defining an [EXAMPLE_FAV_COLOR](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/example/settings.jsonc#L29-L31) environment variable, and [referencing it in the html output file](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/example/example.html#L27). You can reference pod_env_variables in the POD python driver script, too. The ocn_surf_flux POD shows this for several pod_env_vars defined in the [settings.jsonc](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/ocn_surf_flux_diag/settings.jsonc#L16-L21) file that are referenced in the [ocn_surf_flux_diag.py driver script](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/ocn_surf_flux_diag/ocn_surf_flux_diag.py#L64-L65).

wrongkindofdoctor commented:
> @GNellyEmlaw Ah, you haven't committed any of your local changes. Do the following to update your remote branch with changes you made to the files in your POD directory (PR will update automatically):
> 
> ```
> git add diagnostics/TC_rain/*.jsonc
> git add diagnostics/TC_rain/TC_rain.py
> git add diagnostics/TC_rain/TC_rain.html
> ```
> 
> This will add the changes you made to the files in the TC_rain directory without the trailing blank space.
> 
> Now, to remove the duplicate directory with the trailing space from the git commit history, run
> 
> ```
> git rm -r 'diagnostics/TC_rain ' [make sure to add a the space after TC_rain]
> ```
> 
> If you run git status, you should see a message indicating that the changes in the files you just added and removed are staged for a commit. You can now push the changes to your remote branch:
> 
> ```
> git push -u origin TC_rain
> ```
> 
> Once the PR updates and passes the CI, you can move the branch from draft mode to "ready for review". THe framework team can give a more thorough review of the code, and we make suggestions on how to improve the driver script, including modifying some of the parameters so that users can adjust them.
> 
> This is probably an instance where you'd want to incorporate your own POD environment variables ([pod_env_vars](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/main/doc/sphinx/ref_settings.rst#settings-section)), which are set in the settings.jsonc file as you mentioned. The example POD demonstrates this by defining an [EXAMPLE_FAV_COLOR](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/example/settings.jsonc#L29-L31) environment variable, and [referencing it in the html output file](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/example/example.html#L27). You can reference pod_env_variables in the POD python driver script, too. The ocn_surf_flux POD shows this for several pod_env_vars defined in the [settings.jsonc](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/ocn_surf_flux_diag/settings.jsonc#L16-L21) file that are referenced in the [ocn_surf_flux_diag.py driver script](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/40dd5b3adc5dadf45871e7866b5180537324d791/diagnostics/ocn_surf_flux_diag/ocn_surf_flux_diag.py#L64-L65).

@GNellyEmlaw Sorry, I forgot to add the step between the `git add` / `git rm`, and `git push`. After you've added/removed the files to/from the staging area, run:
```
git commit -m "<insert a description of the changes you made here>"
```
The  `-m` is to include the commit description message in the quotes; just running `git commit` will launch a text editor--vim by default on Linux/Unix--with a space for you to add the commit message to the log file. If you go this route, add the message, close the text editor, and continue to the push step.

The changes have now been added to the commit history, and you can push them to your remote branch:
```
git push -u origin TC_rain
```

GNellyEmlaw commented:
@wrongkindofdoctor Got it, thanks for adding that! I was trying to push it last night and it wasn't working, but I was able to push the TC_Rain up to date this morning. I also added the environment variables. Crossing fingers it runs successfully here. 

Thanks so much for your patience and help!

wrongkindofdoctor commented:
@GNellyEmlaw Awesome! Your quick responses and targeted questions have helped immensely in resolving bugs. This thread will likely be useful for other POD developers--particularly those new to GitHub. It will also help the framework team to improve the documentation, discussions threads, and tutorials.

GNellyEmlaw commented:
@wrongkindofdoctor Thank you so much again!

It looks like the branch still had not passed all the checks? In the check details it looks like the checks are stalling at the ocn_surf_flux_diag and temp_extremes_distshape PODS. I don't believe I edited anything within those PODs. Is there anything I can do to fix this?

wrongkindofdoctor commented:
@GNellyEmlaw You need to change the time units back to 'days' in [fieldlist_CMIP.jsonc](https://github.com/GNellyEmlaw/MDTF-diagnostics/blob/f3df49ff154151221150da2c15e4101df0b3b7c1/data/fieldlist_CMIP.jsonc#L12). If the units have to be 'hours' instead of 'days' to run your POD on the test data, let me know, and I will add it to my list of things to fix in the framework. You can keep this change in your local copy of the fieldlist for testing--just exclude fieldlist_CMIP.jsonc (and any other files that you modify that are not in your POD directory) when you are adding+committing+pushing changes to your remote branch.

That said, please also remove changes you made to src/default_tests.jsonc and the [fieldlist_NCAR.jsonc](https://github.com/GNellyEmlaw/MDTF-diagnostics/blob/TC_Rain/data/fieldlist_NCAR.jsonc) from the PR.

GNellyEmlaw commented:
@wrongkindofdoctor ah, I see, thanks! I've removed the changes and pushed the branch up to date, but my POD test data is in hours - sorry to add another to-do to your list!

wrongkindofdoctor commented on src/default_tests.jsonc:
```diff
@@ -105,10 +105,10 @@
 
   // Set to true to have PODs save postscript figures in addition to bitmaps.
-  "save_ps": false,
+  "save_ps": true,
```
```suggestion
  "save_ps": false,
```

wrongkindofdoctor commented on diagnostics/TC_Rain/doc/TC_Rain.rst:
```diff
@@ -0,0 +1,71 @@
+-------------
+
+In the current draft code, the data is loaded in first. The netcdf file for the total 
+precipitatin is loaded in through xarray. The TC track data is read in line by line from a 
```
```suggestion
precipitation is loaded in through xarray. The TC track data is read in line by line from a 
```

wrongkindofdoctor commented on diagnostics/TC_Rain/doc/TC_Rain.rst:
```diff
@@ -0,0 +1,71 @@
+References
+----------
+
+Will concult advisor to find what referene would be good to include here. 
```
Update the references once you've coordinated with your advisor

wrongkindofdoctor commented on diagnostics/TC_Rain/TC_rain.py:
```diff
@@ -0,0 +1,291 @@
+
+# In this example code, the model output is interpolated and regridded to 0.25 x 0.25 
+# degree arrays and the average is calcuated from 0 to 600 km from the center of the 
+# storm in 25 km discrete sections. Here we only take the avearge of snaphots where the
```
```suggestion
# storm in 25 km discrete sections. Here we only take the average of snaphots where the
```

wrongkindofdoctor commented on diagnostics/TC_Rain/TC_rain.py:
```diff
@@ -0,0 +1,291 @@
+    "SP":[[-30,0],[130,210]],
+    "SIO":[[-30,0],[40,130]]}
+
+#get only feild of view for basin storm is in
+
```
```suggestion
#get only field of view for basin storm is in

```

GNellyEmlaw commented:
@wrongkindofdoctor Thank you for the requested changes. I believe I resolved them all except for the references, my advisor is just finishing his sabbatical right now and busy with moving soon so he has not responded to my slack message yet. I meet with him this afternoon so I will definitely get a response by then.

In the mean time I noticed there was a problem with the last check. It looks like there was a problem in generating the model data. I don't believe I changed anything in my last commit, but I might have on accident. Is there anything I can do to resolve this?

GNellyEmlaw commented:
@wrongkindofdoctor Hi Jess, I've resolved all the requested changes. I know from Daehyun that the implementation of this POD is due at the end of July, and I know that is just a few days away. I just want to make sure there is nothing else I need to do, don't want to drop the ball. Sorry to have left everything to the last second.

jkrasting commented on diagnostics/TC_Rain/TC_rain.py:
```diff
@@ -0,0 +1,299 @@
+# 
+#   Version & Contact info
+# 
+#   Here you should describe who contributed to the diagnostic, and who should be
+#   contacted for further information:
```
```suggestion
#   Here you should describe who contributed to the diagnostic, and who should be
#   contacted for further information:
```

jkrasting commented on diagnostics/TC_Rain/TC_rain.py:
```diff
@@ -0,0 +1,299 @@
+# This file is part of the TC Rain Diagnostic POD of the MDTF code package (see mdtf/MDTF-diagnostics/LICENSE.txt)
```
Could you please apply the Black code formatter to your .py scripts?  I appreciate this is not in the guide, but it is very helpful!

https://black.readthedocs.io/en/stable/

jkrasting commented on diagnostics/TC_Rain/doc/TC_Rain.rst:
```diff
@@ -0,0 +1,107 @@
+Climate, Accepted.
+
+
+More about this diagnostic
+--------------------------
```
Would it be possible to add some more text about how a model developer could use this diagnostic?  What should one be looking for in terms of the radial distribution of precip.?  Are there common biases to be on the lookout for?   What would a ideal distribution look like?

jkrasting reviewed (changes requested):
Excellent work @GNellyEmlaw! Thanks!

No major code issues.  Just a few nits.  The only big suggestion is to add some more documentation to help a model developer make use of this POD.

jkrasting reviewed (approved):
Thanks @GNellyEmlaw - looks good!

wrongkindofdoctor commented on diagnostics/TC_Rain/TC_Rain.html:
```diff
@@ -0,0 +1,22 @@
+This POD calculates and plots azimuthal averages for tropical cyclone (TC) rain rates 
+from TC track data and model output precipitation. 
+</p>
+<p>
+The names of any environment variables set for the diagnostic (enclosed in 
+curly braces) will be replaced by the same values they had when the diagnostic 
+was run. For example, this diagnostic's favorite color is "{{EXAMPLE_FAV_COLOR}}."
+</p>
```
```suggestion

```

wrongkindofdoctor commented on diagnostics/TC_Rain/TC_Rain.html:
```diff
@@ -0,0 +1,22 @@
+</p>
+<TABLE>
+<TR>
+<TH align=left><font color=navy>Total Precipiation, {{FIRSTYR}}-{{LASTYR}}
```
```suggestion
<TH align=left><font color=navy>Total Precipitation, {{FIRSTYR}}-{{LASTYR}}
```

wrongkindofdoctor commented on diagnostics/TC_Rain/TC_rain.py:
```diff
@@ -0,0 +1,347 @@
+    "spc": [[-30, 0], [160, 240]],  # south pacific central basin
+}
+
+# get only feild of view for basin storm is in
```
```suggestion
# get only field of view for basin storm is in
```

wrongkindofdoctor commented:
@GNellyEmlaw All set! The TC_rain POD is now part of the MDTF-diagnostics repo, and is ready to use with the main branch. It will appear in the main branch of your fork once you pull in the updates. You can delete your POD development branch, or retain it for reference if you wish. Note that the framework team or other users might make minor changes to the TC_rain POD to solve issues related to things like package version errors, compatibility with framework upgrades, or similar.

GNellyEmlaw commented:
@wrongkindofdoctor Great! Sounds good. Thank you!
